docs/rfcs/20221007-wal-on-kafka.md

```diff
@@ -0,0 +1,106 @@
+| +--------------------+ |        | +-------------------+ |                                                               
+|         ...            |        |          ...          |                                                               
++------------------------+        +-----------------------+                                                               
+
+```
+### 1.2 Record Format
+Wal record, we choose the key with detail location information for using such record format in not only the implementation on kafka.
+
+```
+Key:
++---------------+----------------+-------------------+--------------------+--------------------+
+| namespace(u8) | region_id(u64) |   table_id(u64)   |  sequence_num(u64) | version header(u8) |
++---------------+----------------+-------------------+--------------------+--------------------+
+
+Value:
++--------------------+----------+
+| version header(u8) | playload |
++--------------------+----------+
+```
+### 1.3 Data structure
+The data structure used to record above information is like this (just use to display, thread safe is not considered yet) : 
+
+```
+/// Used to record one table's log state.
+struct LogState {
+    /// The sequence number of the newest log entry 
+    next_sequence_num: SequenceNumber,
+    /// The kafka offset in its partition of the newest log entry 
+    cur_offset: Option<KafkaOffset>,
+    /// The start kafka offset of current log set 
+    /// which is still not deleted(due to compaction)
+    start_offset: Option<KafkaOffset>,
+}
+
+/// All tables in the same shard will be managed by a `Region`.
+struct Region {
+    ...
+    log_states: HashMap<TableId, LogState>,
+    ...
+}
+```
+## 2. Main process
+### 2.1 Open namespace
+- Fetch all topics and filter them according to the namespace prefix. Cache the results after.
+### 2.2 Open region
+- Search the region in opened namespace. 
+- If not found and auto creating is defined,  create corresponding topic in kafka and add it to cache.
+- Return the pointer of `region`.
+### 2.3 Write
+- Open the corresponding `region` (define auto creating).
+- Put the log batch to `wal partition`.
+- Update `cur_offset` and `next_sequence_number` in memory.
+### 2.4 Read
+- Open the corresponding `region`.
+- Just read all the logs of the `region`, the split and replay work will be done on the high level.
+### 2.5 Delete
+Log's deletion can be split to two steps:
++ Mark the delete offset.
++ Do delayed deletion periodically in a cleaner thread.
+#### Mark
++ Each table's `cur_sequence_num` and `cur_offset` will be updated in each log writing, and `start_offset` will be updated in its first log writing.
++ When `mark_delete_entries_up_to` is called by a table, it will make `start_offset` None.
+#### Delete
+The deletion logic done in a cleaner thread, and it may be a bit complicated. In cleaner's checking, I think logs should be considered to delete in these two situation:
```

Comment:
```suggestion
The deletion logic done in a background thread called `cleaner`, and it may be a bit complicated. In the `cleaner`'s checking, two situations should be taken into cosiderations:
```